This snakemake workflow analyzes DNA metabarcoding data with the OBITools v1.2.11 and SUMACLUST.

Sequence analysis was performed using the OBITools v1.2.11 and the SUMACLUST softwares. Forward and reverse reads were aligned with the *illuminapairedend* program (rule **pairing**). Alignments with a score below or equal to {{ snakemake.config["alifilt"]["minscore"] }} were filtered out. Sequences were then assigned to samples using the *ngsfilter* command, allowing 0 and 2 errors on the tags and primers respectively (rule **demultiplex**). Identical sequences were merged with the *obiuniq* command (rule **derepl**). Next, low quality sequences were filtered out, i.e. sequences with (or less than) {{ snakemake.config["basicfilt"]["mincount"] }} reads in the whole dataset, sequences containing ambiguous bases or that are shorter than expected (<{{ snakemake.config["basicfilt"]["minlength"] }} bp) (rule **basicfilt**). The remaining sequences were clustered into Molecular Operational Taxonomic Units (MOTUs) with *sumaclust* using a similarity threshold of {{ snakemake.config["clustering"]["minsim"] }} (rule **clustering**). The most abundant sequence of each MOTU was considered as its representative sequence and its abundance corresponded to the sum of the abundance of its members. The fasta file resulting from these analysis steps was converted to a tabular file containing MOTUs abundances in samples (rule **tab_format**).